File: Airship/AirshipCore/Source/Internal/UARemoteDataAPIClient.m
/* Copyright Airship and Contributors */

#import "UARemoteDataAPIClient+Internal.h"
#import "UAUtils+Internal.h"
#import "UARuntimeConfig.h"
#import "UAirshipVersion.h"
#import "UAirship.h"
#import "NSError+UAAdditions.h"

@interface UARemoteDataResponse()
@property (nonatomic, copy, nullable) NSArray<NSDictionary *> *payloads;
@property (nonatomic, copy, nullable) NSString *lastModified;
@end

@implementation UARemoteDataResponse
- (instancetype)initWithStatus:(NSUInteger)status
                      payloads:(NSArray<NSDictionary *> *)payloads
                  lastModified:(NSString *)lastModified {
    self = [super initWithStatus:status];
    if (self) {
        self.payloads = payloads;
        self.lastModified = lastModified;
    }
    return self;
}
@end


@interface UARemoteDataAPIClient()
@property (nonatomic, strong) UARequestSession *session;
@property (nonatomic, strong) UARuntimeConfig *config;
@end

@implementation UARemoteDataAPIClient

NSString * const UARemoteDataAPIClientPath = @"api/remote-data/app";
NSString * const UARemoteDataAPIClientErrorDomain = @"com.urbanairship.remote_data_api_client";

- (UARemoteDataAPIClient *)initWithConfig:(UARuntimeConfig *)config
                                  session:(UARequestSession *)session {
    self = [super init];

    if (self) {
        self.config = config;
        self.session = session;
    }

    return self;
}

+ (UARemoteDataAPIClient *)clientWithConfig:(UARuntimeConfig *)config
                                    session:(UARequestSession *)session {
    return [[self alloc] initWithConfig:config
                                session:session];
}

+ (UARemoteDataAPIClient *)clientWithConfig:(UARuntimeConfig *)config {
    return [[self alloc] initWithConfig:config
                                session:[UARequestSession sessionWithConfig:config]];
}


- (UADisposable *)fetchRemoteDataWithLocale:(NSLocale *)locale
                               lastModified:(nullable NSString *)lastModified
                          completionHandler:(UARemoteDataAPIClientCompletionHandler)completionHandler {

    NSURL *URL = [UARemoteDataAPIClient createRemoteDataURLWithURL:self.config.remoteDataAPIURL
                                                            appKey:self.config.appKey
                                                            locale:locale];

    UARequest *request = [UARequest requestWithBuilderBlock:^(UARequestBuilder * _Nonnull builder) {
        builder.URL = URL;
        builder.method = @"GET";
        [builder setValue:lastModified forHeader:@"If-Modified-Since"];
    }];

    UA_LTRACE(@"Request to update remote data: %@", URL);
    return [self.session performHTTPRequest:request
                          completionHandler:^(NSData * _Nullable data, NSHTTPURLResponse * _Nullable response, NSError * _Nullable error) {

        UA_LTRACE(@"Fetch finished with response: %@ error: %@", response, error);

        if (error) {
            completionHandler(nil, error);
            return;
        }

        if (response.statusCode == 200) {
            NSArray *payloads = [self parseRemoteData:data error:&error];

            if (error) {
                UA_LTRACE(@"Failed to parse remote data with error: %@", error);
                completionHandler(nil, error);
            } else {
                NSString *lastModified = [response.allHeaderFields objectForKey:@"Last-Modified"];
                UARemoteDataResponse *remoteDataResponse = [[UARemoteDataResponse alloc] initWithStatus:response.statusCode
                                                                                               payloads:payloads lastModified:lastModified];

                completionHandler(remoteDataResponse, nil);
            }
        } else {
            UARemoteDataResponse *remoteDataResponse = [[UARemoteDataResponse alloc] initWithStatus:response.statusCode
                                                                                           payloads:nil
                                                                                       lastModified:nil];
            completionHandler(remoteDataResponse, nil);
        }
    }];
}

- (NSArray *)parseRemoteData:(nullable NSData *)data
                       error:(NSError **)error {
    // Missing response body
    if (!data) {
        *error = [NSError airshipParseErrorWithMessage:@"Refresh remote data missing response body."];
        return nil;
    }

    // Parse the response
    NSError *parseError;
    NSDictionary *jsonResponse = [NSJSONSerialization JSONObjectWithData:data
                                                                 options:NSJSONReadingAllowFragments error:&parseError];


    if (!parseError) {
        return [jsonResponse objectForKey:@"payloads"];
    } else {
        *error = parseError;
        return nil;
    }
}

+ (NSURL *)createRemoteDataURLWithURL:(NSString *)remoteDataAPIURL
                               appKey:(NSString *)appKey
                               locale:(NSLocale *)locale {

    NSURLQueryItem *languageItem = [NSURLQueryItem queryItemWithName:@"language"
                                                               value:[locale objectForKey:NSLocaleLanguageCode]];
    NSURLQueryItem *countryItem = [NSURLQueryItem queryItemWithName:@"country"
                                                              value:[locale objectForKey:NSLocaleCountryCode]];
    NSURLQueryItem *versionItem = [NSURLQueryItem queryItemWithName:@"sdk_version"
                                                              value:[UAirshipVersion get]];

    NSURLComponents *components = [NSURLComponents componentsWithString:remoteDataAPIURL];

    // api/remote-data/app/{appkey}/{platform}?sdk_version={version}&language={language}&country={country}
    components.path = [NSString stringWithFormat:@"/%@/%@/%@", UARemoteDataAPIClientPath, appKey, @"ios"];

    NSMutableArray *queryItems = [NSMutableArray arrayWithObject:versionItem];

    if (languageItem.value != nil && languageItem.value.length != 0) {
        [queryItems addObject:languageItem];
    }

    if (countryItem.value != nil && countryItem.value.length != 0) {
        [queryItems addObject:countryItem];
    }

    components.queryItems = queryItems;

    return [components URL];
}

@end

File: Airship/AirshipCore/Source/UARemoteDataAPIClient.swift
/* Copyright Airship and Contributors */

/**
 * @note For internal use only. :nodoc:
 */
@objc
public class UARemoteDataAPIClient : NSObject {
    private let path = "api/remote-data/app"
    private let session: UARequestSession
    private let config: UARuntimeConfig

    @objc
    public init(config: UARuntimeConfig, session: UARequestSession) {
        self.config = config
        self.session = session
        super.init()
    }

    @objc
    public convenience init(config: UARuntimeConfig) {
        self.init(config: config, session: UARequestSession(config: config))
    }

    @objc
    @discardableResult
    public func fetchRemoteData(withLocale locale: Locale, lastModified: String?, completionHandler: @escaping (UARemoteDataResponse?, Error?) -> Void) -> UADisposable {
        let url = remoteDataURL(withLocale: locale)
        let request = UARequest(builderBlock: { builder in
            builder.url = url
            builder.method = "GET"
            builder.setValue(lastModified, header: "If-Modified-Since")
        })

        AirshipLogger.debug("Request to update remote data: \(request)")

        return session.performHTTPRequest(request, completionHandler: { (data, response, error) in

            guard let response = response else {
                AirshipLogger.debug("Fetch finished with \(error?.localizedDescription ?? "")")
                completionHandler(nil, error)
                return
            }

            if response.statusCode == 200 {
                do {
                    let payloads = try self.parseRemoteData(data)
                    let lastModified = response.allHeaderFields["Last-Modified"] as? String
                    let remoteDataResponse = UARemoteDataResponse(
                        status: 200,
                        requestURL: url,
                        payloads: payloads,
                        lastModified: lastModified)
                    completionHandler(remoteDataResponse, nil)
                } catch  {
                    AirshipLogger.debug("Failed to parse remote data with error: \(error)")
                    completionHandler(nil, error)
                    return
                }
            } else {
                let remoteDataResponse = UARemoteDataResponse(
                    status: response.statusCode,
                    requestURL: url,
                    payloads: nil,
                    lastModified: nil)
                completionHandler(remoteDataResponse, nil)
            }
        })
    }

    func parseRemoteData(_ data: Data?) throws -> [AnyHashable]? {
        if data == nil {
            throw AirshipErrors.parseError("Refresh remote data missing response body.")
        }

        // Parse the response
        let jsonResponse = try JSONSerialization.jsonObject(with: data!, options: .allowFragments) as? [AnyHashable : Any]
        return jsonResponse?["payloads"] as? [AnyHashable]
    }

    @objc
    public func remoteDataURL(withLocale locale: Locale) -> URL? {
        let languageItem = URLQueryItem(name: "language", value: locale.languageCode)
        let countryItem = URLQueryItem(name: "country", value: locale.regionCode)
        let versionItem = URLQueryItem(name: "sdk_version", value: UAirshipVersion.get())

        var components = URLComponents(string: config.remoteDataAPIURL ?? "")

        // api/remote-data/app/{appkey}/{platform}?sdk_version={version}&language={language}&country={country}
        components?.path = "/\(path)/\(config.appKey)/\("ios")"

        var queryItems = [versionItem]

        if languageItem.value != nil && (languageItem.value?.count ?? 0) != 0 {
            queryItems.append(languageItem)
        }

        if countryItem.value != nil && (countryItem.value?.count ?? 0) != 0 {
            queryItems.append(countryItem)
        }

        components?.queryItems = queryItems as [URLQueryItem]
        return components?.url
    }
}
